### PR TITLE
Update how table uses are reported

### PR DESCRIPTION
**Pull request by systay**
## Description
Currently, table uses reporting is done using the `Table` and `Keyspace` fields on the `LogStats` struct.

Unfortunately, queries are often contain multiple tables, and even tables from different keyspaces. The current form of reporting this information is not very clean, and it doesn't help you understand which tables exist in which keyspaces.

In this PR, I want to change how this is reported, using a new field in `LogStats`. It will be of type `[]string` and contain the tables used in the query, in the format: `keyspace.table`. When no keyspace exists (e.g. `dual`), just the table name will be returned.

This should make it easier to keep track of table usages going forward.

## Related Issue(s)
#10536

### CODE REVIEW

**Comment by DeathBorn** [comment]
Prometheus metrics  `vtgate_queries_processed_by_table` , `vtgate_queries_routed_by_table` also have keyspace and table lables confusing. In a way, one can see multi table label as monitoring for joins. However, only 1 keyspace is reported.

File: go/vt/vtgate/planbuilder/builder.go [hunk 1]
```diff
@@ -403,7 +452,51 @@ func buildFlushTables(stmt *sqlparser.Flush, vschema plancontext.VSchema) (engin
...
+	tables map[string]any
+}
+
+func (tc *tableCollector) addTable(ks, tbl string) {
+	if tc.tables == nil {
+		tc.tables = map[string]any{}
+	}
+	tc.tables[fmt.Sprintf("%s.%s", ks, tbl)] = nil
+}
+func (tc *tableCollector) addASTTable(ks string, tbl sqlparser.TableName) {
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/builder.go:468]: nit: let's add a new line

**Review by frouioui — APPROVED** [review]
me gusta. left a few questions and nits

File: go/vt/vtgate/logstats/logstats.go [hunk 2]
```diff
@@ -71,10 +73,9 @@ func NewLogStats(ctx context.Context, methodName, sql, sessionUUID string, bindV
 	}
 }
 
-// Send finalizes a record and sends it
-func (stats *LogStats) Send() {
+// SaveEndTime sets the end time of this request to now
+func (stats *LogStats) SaveEndTime() {
 	stats.EndTime = time.Now()
-	QueryLogger.Send(stats)
 }
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/logstats/logstats.go:79]: I wanted to import as little as possible to the logstats package, that's why I split this up.

File: go/vt/vtgate/logstats/logstats.go [hunk 3]
```diff
@@ -152,36 +153,78 @@ func (stats *LogStats) Logf(w io.Writer, params url.Values) error {
 	var fmtString string
 	switch *streamlog.QueryLogFormat {
 	case streamlog.QueryLogFormatText:
-		fmtString = "%v\t%v\t%v\t'%v'\t'%v'\t%v\t%v\t%.6f\t%.6f\t%.6f\t%.6f\t%v\t%q\t%v\t%v\t%v\t%q\t%q\t%q\t%q\t%q\t%v\t\n"
+		if len(stats.TablesUsed) == 0 {
+			fmtString = "%v\t%v\t%v\t'%v'\t'%v'\t%v\t%v\t%.6f\t%.6f\t%.6f\t%.6f\t%v\t%q\t%v\t%v\t%v\t%q\t%q\t%q\t%q\t%q\t%v\n"
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/logstats/logstats.go]: 👍

File: go/vt/vtgate/planbuilder/builder.go [hunk 4]
```diff
@@ -75,15 +112,23 @@ func TestBuilder(query string, vschema plancontext.VSchema, keyspace string) (*e
...
+		primitive = planResult.primitive
+		tablesUsed = planResult.tables
+	}
 	plan := &engine.Plan{
 		Type:         sqlparser.ASTToStatementType(stmt),
 		Original:     query,
-		Instructions: instruction,
+		Instructions: primitive,
 		BindVarNeeds: bindVarNeeds,
+		TablesUsed:   tablesUsed,
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/builder.go:131]: planResult is sometimes nil, so we need to do the nil-check:

```go
	if planResult != nil {
		primitive = planResult.primitive
		tablesUsed = planResult.tables
	}
```

File: go/vt/vtgate/planbuilder/gen4_compare_v3_planner.go [hunk 5]
```diff
@@ -64,11 +63,13 @@ func gen4CompareV3Planner(query string) func(sqlparser.Statement, *sqlparser.Res
...
-			V3:         v3Primitive,
-			Gen4:       gen4Primitive,
+		primitive := &engine.Gen4CompareV3{
+			V3:         v3Primitive.primitive,
+			Gen4:       gen4Primitive.primitive,
 			HasOrderBy: hasOrderBy,
-		}, nil
+		}
+
+		return newPlanResult(primitive, gen4Primitive.tables...), nil
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/gen4_compare_v3_planner.go:72]: yep

**Review by DeathBorn — APPROVED** [review]
Looks nice!

File: go/vt/vtgate/logstats/logstats.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (frouioui) [reply go/vt/vtgate/logstats/logstats.go:79]: Got it

[the remark this answers is not in the document]
Comment (harshit-gangal) [reply go/vt/vtgate/logstats/logstats.go:79]: I do not see any method in logstats expect send. Now, as we are not seding the logstats here. We can set the EndTime outside itself like we do for other fields of logStats. why to do EndTime set differently.

File: go/vt/vtgate/executor.go [hunk 6]
```diff
@@ -314,6 +317,7 @@ func (e *Executor) StreamExecute(
 		// 5: Log and add statistics
 		logStats.Keyspace = plan.Instructions.GetKeyspaceName()
 		logStats.Table = plan.Instructions.GetTableName()
+		logStats.TablesUsed = plan.TablesUsed
```
Comment (harshit-gangal) [inline go/vt/vtgate/executor.go:320]: Is keyspace and table are still kept in logStats for backward compatibility?
If so, we should mark it for deprecation

File: go/vt/vtgate/planbuilder/testdata/lock_cases.txt [hunk 7]
```diff
@@ -16,7 +16,25 @@
...
+    "TargetDestination": "KeyspaceID(00)",
+    "FieldQuery": "select get_lock('xyz', 10) from dual where 1 != 1",
+    "lock_func": [
+      "get_lock('xyz', 10)"
+    ]
+  },
+  "TablesUsed": [
+    "dual"
+  ]
+}
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/lock_cases.txt:37]: other places it is <ks>.dual and in lock_cases it is only "dual", why this is different?

**Comment by harshit-gangal** [comment]
Added benchmark label to see if there is any change in numbers.

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vtgate/planbuilder/testdata/lock_cases.txt (hunk 7 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/lock_cases.txt]: found and fixed

**Comment by harshit-gangal** [comment]
added `release notes` label to add deprecation warning in summary release notes.

**Pull request merged**